### PR TITLE
Added toQuaternion and fromQuaternion methods

### PR DESCRIPTION
**Pull request by francesco-romano**
cc @traversaro

### CODE REVIEW

File: src/core/include/iDynTree/Core/Rotation.h [hunk 1]
```diff
@@ -157,6 +157,31 @@ namespace iDynTree
...
+        bool getQuaternion(iDynTree::Vector4& quaternion) const;
+
+        /**
+         * Get the corresponding quaternion components for this rotation
+         *
+         * The quaternion is defined as [s, r]
+         * where s \in \mathbb{R} is the costituent and
+         * r \in \mathbb{R}^3 is the imaginary part.
+         *
+         * @param[out] s the imaginary part
```
Comment (traversaro) [inline src/core/include/iDynTree/Core/Rotation.h]: the real part

Comment (francesco-romano) [inline src/core/include/iDynTree/Core/Rotation.h]: 😆

File: src/core/include/iDynTree/Core/Rotation.h [hunk 2]
```diff
@@ -271,6 +296,19 @@ namespace iDynTree
...
+         *
+         * The quaternion is expected to be ordered in the following way:
+         * - s \in \mathbb{R} the constituent of the quaterion
+         * - r \in \mathbb{R}^3 the imaginary part of the quaternion
+         *
+         * @param quaternion a quaternion representing a rotation
+         *
+         * @return The rotation matrix
+         */
+        static Rotation FromQuaternion(const iDynTree::Vector4& quaternion);
```
Comment (traversaro) [inline src/core/include/iDynTree/Core/Rotation.h]: Most methods (especially `from...` methods) start with a lower case letter.

File: src/core/src/Rotation.cpp [hunk 3]
```diff
@@ -286,6 +286,71 @@ namespace iDynTree
...
+            q1 *= (R(2,1) + R(1,2)) >= 0 ? 1 : -1;
+            q2 *= 1.0;
+            q3 *= (R(3,2) + R(2,3)) >= 0 ? 1 : -1;
+        } else if (q3 >= q0 && q3 >= q1 && q3 >= q2) {
+            q0 *= (R(2,1) - R(1,2)) >= 0 ? 1 : -1;
+            q1 *= (R(1,1) + R(1,3)) >= 0 ? 1 : -1;
+            q2 *= (R(3,2) + R(2,3)) >= 0 ? 1 : -1;
+            q3 *= 1.0;
+        } else {
+            std::cerr << "Quaternion numerically bad conditioned\n";
```
Comment (traversaro) [inline src/core/src/Rotation.cpp]: std::cerr << "Rotation::getQuaternion : Quaternion numerically bad conditioned\n"; or use `reportError` function.

File: src/core/include/iDynTree/Core/Rotation.h [hunk 4]
```diff
@@ -271,6 +296,19 @@ namespace iDynTree
          */
         static Rotation Identity();
 
+        /**
+         * Construct a rotation matrix from the given quaternion representation
+         *
+         * The quaternion is expected to be ordered in the following way:
+         * - s \in \mathbb{R} the constituent of the quaterion
```
Comment (traversaro) [inline src/core/include/iDynTree/Core/Rotation.h]: I imagine constitutent is another name for the real part, right?

File: src/core/include/iDynTree/Core/Rotation.h (hunk 2 above)
Comment (francesco-romano) [inline src/core/include/iDynTree/Core/Rotation.h]: Yes, but all yours static methods start with capital letter.
I thought it was a convention for iDynTree.

Happy to change it to lower.
Should I?

File: src/core/include/iDynTree/Core/Rotation.h (hunk 4 above)
Comment (francesco-romano) [inline src/core/include/iDynTree/Core/Rotation.h]: It's written in the reference where I took the formula.
Should I change it to real part?

File: src/core/include/iDynTree/Core/Rotation.h (hunk 2 above)
Comment (traversaro) [inline src/core/include/iDynTree/Core/Rotation.h]: I just noticed that this is a static method. Now I see, even if in the existing objects we tend to have actual methods called `from...` that modify the `*this` object.

File: src/core/src/Rotation.cpp (hunk 3 above)
Comment (francesco-romano) [inline src/core/src/Rotation.cpp]: 👍

**Comment by traversaro** [comment]
Not necessary to add it now, but for bindings use I think that a:
~~~
iDynTree::Vector4 Rotation::asQuaternion()
~~~
would be pratically really useful.

File: src/core/include/iDynTree/Core/Rotation.h (hunk 2 above)
Comment (francesco-romano) [inline src/core/include/iDynTree/Core/Rotation.h]: should we find another name?
I don't like `Quaternion` as it is not clear that we are creating a rotation matrix from a quaternion.

`Rotation::RotationFromQuaternion` ?

**Comment by francesco-romano** [comment]
@traversaro `asQuaternion` added in 01c0224021bde1beded740f48e2400f384da90b9

**Comment by traversaro** [comment]
I have added a (failing) test in https://github.com/robotology/idyntree/pull/167/commits/c4a3f1d7cc78a81c62fa592e157fbfc3f47be034 . The test passes because I commented out the actual checks, but it is possible to see how the results do not match the one in YARP and KDL in a simple case. You can compile and run the test enabling the option `IDYNTREE_COMPILE_TESTS` and by running:
~~~
ctest -VV -R Quaternion
~~~

**Comment by francesco-romano** [comment]
@traversaro I'll take care of it

**Comment by traversaro** [comment]
Everything working fine, merging!

**Pull request merged**